### ISSUE TITLE
Is the code badly written, or this is how Rust actually supposed to look like?

### ISSUE THREAD

**Issue by odiferousmint** [uutils/coreutils#1342]
https://github.com/uutils/coreutils/blob/master/src/echo/echo.rs

This looks horrid. Is it because you need to workaround the compiler's limitations in the name of safety, or the author doesn't know Rust, or what is the reason for it being so overly complicated and bloated?

```
usize::from_str_radix(from_utf8(bytes.as_ref()).unwrap(), base as u32).unwrap() as u8 as char
```

50% type conversion/cast
49% discarding the error
1% actual code.

Why?

C version: https://github.com/openbsd/src/blob/master/bin/echo/echo.c

(Yes, the features are not identical, but it actually wouldn't be too many more lines to implement the same functionality in the C version.)

---

https://github.com/uutils/coreutils/blob/master/src/cp/cp.rs

vs.

https://github.com/openbsd/src/blob/master/bin/cp/cp.c

Ugh.

A fair comparison: https://github.com/coreutils/coreutils/blob/master/src/cp.c

GNU is known to be bloated, and it is a few lines longer than the Rust's one, but C has manual memory management, and Rust is supposed to be a higher-level language anyway.

---

```
let src_path = CString::new(source.as_os_str().to_str().unwrap()).unwrap();
hard_links.push((dest.clone().to_str().unwrap().to_string(), inode));
```

Ugh... is this efficient? Rust has been sold as having "zero cost abstractions", so not sure why this is a thing. I do not even blame the developers, when I tried to write Rust code, I had to do stuff like this, too.

---

I don't mean to be hostile, I am just baffled at why people love Rust so much. All I see is undelivered promises.

**Comment by Arcterus** [comment]
1. A lot of the code is very old (like pre-1.0, so both the language and library have changed significantly), meaning code that previously looked fine now has various quick fixes to get the code working again.
2. New Rust programmers frequently contribute code, so the code is not always as clean as possible.
3. GNU’s functionality is sometimes semi-convoluted to implement cleanly.
4. Supporting Unix (with special code for just macOS, Linux, etc. sometimes), Windows, and Redox can necessitate more complicated code.  OpenBSD’s code doesn’t have to deal with that, and (AFAIK) GNU just shoves most things into a giant library trying to make everything act like one platform.
4. Looking at that line in `echo`, errors are ignored because the previous code ensures `bytes` contains only valid character representations of digits, and `base` needs to be cast to `u32` because `Base` is an `enum`.  The other two casts could probably be written differently.
5. The C version implements none of the `match` and only one option, so it makes sense that it is significantly simpler (it would be pretty much just as simple in Rust).

I’m not going to compare the actual GNU coreutils code due to licensing, but I assume the issues are similar with `cp`.

**Comment by rockdaboot** [comment]
Rust is evolving fastly, like C++. So any rust code base bit-rots pretty fast and needs a (partially) rewrite every now and then. Nobody likes to do that kind of work - writing new code is much more fun. And introducing regressions during a re-work dirties your white vest. Look at long-standing C++ projects to see the same.

**Comment by RazrFalcon** [comment]
>49% discarding the error

In this case, it's guarantee that `from_str_radix` and `from_utf8` will not fail. So it's a bad code/API.

**Comment by ccbrown** [comment]
> Nobody likes to do that kind of work - writing new code is much more fun.

I actually love re-inventing the wheel. :)

Threw together a PR with my attempt at a cleaner echo command.

**Comment by kneitinger** [comment]
> I actually love that kind of work. :)

I'd like to echo (I swear no pun intended) what @ccbrown said. I love system utilities programming and am a moderately new rust user who would benefit greatly from such work. Is there anyway we could get a list of utils that could use some attention? Thanks!

**Comment by BurntSushi** [comment]
@kneitinger `grep` appears to be missing, but that's a pretty significant project. I'd be happy to help mentor anyone who wanted to take it on though. (If you were OK giving up on leftmost-longest POSIX regex semantics and maybe some other small semantics on regexes, then it's probably a somewhat doable project.)

**Comment by Arcterus** [comment]
I'll see if I can compile a list of the utils that need the most work in the next week or two.  In terms of `grep`, it's technically not part of GNU coreutils, so I'd be more inclined to make it a separate repository (although we'd probably need more maintainers were I to do so).

**Comment by BurntSushi** [comment]
> it's technically not part of GNU coreutils

TIL. Thanks for the clarification!

**Comment by Arcterus** [comment]
Closed in favor of #1358.
